index/scorch/merge.go

```diff
@@ -129,8 +147,79 @@ func (s *Scorch) parseMergePlannerOptions() (*mergeplan.MergePlanOptions,
 	return &mergePlannerOptions, nil
 }
 
+type mergerCtrl struct {
+	creator string
+	options *mergeplan.MergePlanOptions
+	doneCh  chan struct{}
+}
+
+func (s *Scorch) RequestMerge(mo *mergeplan.MergePlanOptions,
+	doneCh chan struct{}) error {
+	// check whether force merge is already under processing
+	if atomic.LoadUint64(&s.stats.TotFileMergeForceOpsStarted) >
+		atomic.LoadUint64(&s.stats.TotFileMergeForceOpsCompleted) {
+		if doneCh != nil {
+			close(doneCh)
+		}
+		return fmt.Errorf("compaction already in progress")
+	}
+
+	go func() {
+		defer func() {
+			// doneCh for the caller to track the async merge completion
+			if doneCh != nil {
+				close(doneCh)
+			}
+		}()
+
+		if atomic.LoadUint64(&s.stats.TotFileMergeForceOpsStarted) >
+			atomic.LoadUint64(&s.stats.TotFileMergeForceOpsCompleted) {
+			return
+		}
+		atomic.AddUint64(&s.stats.TotFileMergeForceOpsStarted, 1)
```

Comment:
it seems to me that this use of atomics to avoid running multiple force merges isn't safe.  why couldn't 2 competing callers have both run through all the checks, and both gotten to this point, where 2 merges now start?

Reply:
true, you are right!